### PR TITLE
Debug xml dsig

### PR DESCRIPTION
**Pull request by fisx**
These changes make XML signature validation more robust, and also more correct.

Tasks:

- [x] capture stdout, stderr noise from libxml2 and feed it into an error that is returned;
- [x] make sure that the signed subtrees are self-contained by applying an extra round of canonicalization;
- [x] fix test suite (currently: `"failed to canonicalize input: xmlReadMemory: illegal operation (Inappropriate ioctl for device)"`)

### CODE REVIEW

File: SAML2/XML/Signature.hs [hunk 1]
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
...
-  when (null rl) $
-    throwError . SignatureVerifyNoReferences $ show rl
-  unless (all isRight rl) $
-    throwError . SignatureVerifyBadReferences $ show (signedInfoReference si, rl)
-  unless (elem (Right xid) rl) $
-    throwError . SignatureVerifyInputNotReferenced $ show rl
+  signedSubtree :: HXT.XmlTree
+    <- do
+        mdoc' <- liftIO . try $ fixNamespaces doc
+        doc' <- case mdoc' of
```
Comment (ChrisPenner) [inline SAML2/XML/Signature.hs]: This case can be simplified to `fromRight`

File: SAML2/XML/Signature.hs [hunk 2]
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
...
+          Right x
+            -> pure x
+          Left (err :: SomeException)
```
Comment (ChrisPenner) [inline SAML2/XML/Signature.hs]: It'd be nice to move the `SomeException` type annotation up into `try @SomeException` so the exception type stays close to the `try` and makes it clear that we plan to catch ALL exceptions 😄

File: SAML2/XML/Signature.hs [hunk 3]
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
...
+          <- let child n = HXT.runLA $ HXT.getChildren HXT.>>> isDSElem n HXT.>>> HXT.cleanupNamespaces HXT.collectPrefixUriPairs
+             in case child "Signature" signedSubtree of
+            [sx] -> return sx
+            _ -> throwError SignatureNotFoundOrEmpty
+
+        case docToSAML sx of
+          Left err -> throwError . SignatureParseError $ show err
+          Right v -> pure v
+
+  signedInfoElem :: BS.ByteString
```
Comment (ChrisPenner) [inline SAML2/XML/Signature.hs]: Some docstring or comments explaining the magic xPaths would be nice 😄

File: SAML2/XML/Signature.hs [hunk 4]
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
...
+          . capture' "applyCanonicalization"
+          $ (applyCanonicalization (signedInfoCanonicalizationMethod signedInfoTyped) (Just xpath) $ DOM.mkRoot [] [signedSubtree])
+
+  referenceChecks :: NonEmpty.NonEmpty (Either String String)
+    <- failWith (SignatureVerifyReferenceError . (show (signedInfoReference signedInfoTyped) <>))
+      . capture' "verifyReference"
+      $ mapM (`verifyReference` signedSubtree) (signedInfoReference signedInfoTyped)
+
+  unless (all isRight referenceChecks) $
+    throwError . SignatureVerifyBadReferences $ show (signedInfoReference signedInfoTyped, referenceChecks)
```
Comment (ChrisPenner) [inline SAML2/XML/Signature.hs]: If there's a left in the referenceChecks should we possibly `sequence` it to collect the error out? Or is it more descriptive to print everything out like this?

File: SAML2/XML/Signature.hs [hunk 5]
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
...
+  can :: SBS
+    <- liftIO . capture' "fixNamespaces" $
+       canonicalize (CanonicalXMLExcl10 False) Nothing Nothing $ DOM.mkRoot [] [doc]
+  maybe (throwIO . ErrorCall $ "parse error on canonicalized xml") pure $
+    xmlToDoc (cs can)
+
+capture' :: String -> IO a -> IO a
+capture' actionName action = hCapture [stdout, stderr] action >>= \case
+  ("", !out) -> pure out
+  (noise, _) -> throwIO . ErrorCall $ actionName <> ": " <> noise
```
Comment (ChrisPenner) [inline SAML2/XML/Signature.hs]: If any noise on stdout indicates an error; it's probably helpful to have the output of stderr appended here too.

Comment (fisx) [reply SAML2/XML/Signature.hs]: i think that's what this does, no?

(thanks for the other comments, more changes coming up!)

File: SAML2/XML/Signature.hs (hunk 3 above)
Comment (fisx) [reply SAML2/XML/Signature.hs]: Agreed, but i would have to figure out what that code does first...  (It's not mine)

File: SAML2/XML/Signature.hs (hunk 4 above)
Comment (fisx) [reply SAML2/XML/Signature.hs]: It's a list of `Either`s, and I want to keep track of how many errors there are, and which precisely.

I agree it looks weird, but I can't think of a better way to do it.  Wanna try something and add a commit?

File: SAML2/XML/Signature.hs (hunk 5 above)
Comment (ChrisPenner) [reply SAML2/XML/Signature.hs]: Ahh; I see! I misunderstood; It looked at a glance that the tuple matched up with `[stdout, stderr]`; but in hindsight that does't make sense 👍

**Comment by ChrisPenner** [comment]
It's all greek to me 🤷‍♂️ 

Did this not change any behaviour in any of the tests? If that's the case we should almost certainly add some new tests to capture whatever was wrong before this change.

File: SAML2/XML/Signature.hs (hunk 4 above)
Comment (ChrisPenner) [reply SAML2/XML/Signature.hs]: In that case; how about `show (signedInfoReference signedInfoTyped, lefts referenceChecks)` instead? `lefts` will just grab all the errors from the list of Eithers

**Comment by fisx** [comment]
> It's all greek to me 
> 
> Did this not change any behaviour in any of the tests? If that's the case we should almost certainly add some new tests to capture whatever was wrong before this change.

It does make one test fail.  See PR description.  :-)

To reproduce, run `stack test --fast` and follow the error.

**Comment by fisx** [comment]
oh, and:

> It's all greek to me

i've stared at this for over a day now, and not for the first time, and i feel the same way.  :-).  i don't think there is a way to implement xml:dsig though that would make the reader of the code feel any better about the world.  that's just how xml standards are.

**Comment by fisx** [comment]
> `xmlReadMemory: illegal operation (Inappropriate ioctl for device)`

i think i know what's going on here.

Since libxml2 doesn't return errors, but spills them on stdout, stderr, I use [hCapture](http://hackage.haskell.org/package/silently-1.2.5/docs/System-IO-Silently.html) that reads those into a string and returns it with together with the value.

`hCapture` seems to change the type of the file handle (from what to what?).  libxml2 not only spills stuff on stderr, but also performs a ioctl operation on stderr (or stdout) that is supported before `hCapture`, but not after.  hence, [ENOTTY](https://www.gnu.org/software/libc/manual/html_node/Error-Codes.html).

The first solution I can think of is to trap and ignore this error `c14n`.

**Comment by fisx** [comment]
ready for review!

File: SAML2/XML.hs [hunk 6]
```diff
@@ -100,13 +101,24 @@ samlToDoc = head
   . HXT.runLA (HXT.processChildren $ HXT.cleanupNamespaces HXT.collectPrefixUriPairs)
   . XP.pickleDoc XP.xpickle
 
+-- | 'samlToDoc' adds bogus `</>` elements around the output tree, this variant doesn't.
+--
+-- TODO: change this function to return 'HXT.XmlTrees' and replace 'samlToDoc' with it.
+samlToDoc' :: XP.XmlPickler a => a -> HXT.XmlTree
+samlToDoc' = head . getChildren . head
```
Comment (neongreen) [inline SAML2/XML.hs:108]: `samlToDoc' = head . getChildren . samlToDoc` ?

Comment (neongreen) [reply SAML2/XML.hs:108]: Or do you want it to remain like it is so that it would be easier to get rid of `samlToDoc` in the future?

File: test/XML/Signature.hs [hunk 7]
```diff
@@ -302,11 +306,6 @@ examples = zipWith ($) xs [1..]
...
...TEWMBQGA1UEBwwNU2FuIEZyYW5jaXNjbzENMAsGA1UECgwET2t0YTEUMBIGA1UECwwLU1NPUHJvdmlkZXIxEzARBgNVBAMMCmRldi02MDc2NDgxHDAaBgkqhkiG9w0BCQEWDWluZm9Ab2t0YS5jb20wHhcNMTgwNTIzMDg1MTA1WhcNMjgwNTIzMDg1MjA1WjCBkjELMAkGA1UEBhMCVVMxEzARBgNVBAgMCkNhbGlmb3JuaWExFjAUBgNVBAcMDVNhbiBGcmFuY2lzY28xDTALBgNVBAoMBE9rdGExFDASBgNVBAsMC1NTT1Byb3ZpZGVyMRMwEQYDVQQDDApkZXYtNjA3NjQ4MRwwGgYJKoZIhvcNAQkBFg1pbmZvQG9rdGEuY29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA2HkpOuMhVFUCptrVB/Zm36cuFM+YMQjKdtqEoBJDLbtSbb7uFuvm5rMJ+1VSK5GKAM/Bec5WXTE2WMkifK5JaGOLS7q8+pgiWmqKE3KHMUmLAioe/1jzHkCobxis0FIVhyarRY97w0VMbDGzhPiU7pEopYpicJBzRL2UrzR+PebGgllvnaPzlg8ePtr9/xMv0QTJlYEyCctO4vT5Qa5Xlfek3Ox5yMJM1JPXzn7yuJN5R/Nf8jFprsdBSxNMzkcTRFGy8as2GCt/Xh9H+ef4CxSgRK5UXcUCrb5YMnBehEp2YiuWtw8QsGRR8elgnF3Uw9J2xEDkZIhurPy8OYmGNQIDAQABMA0GCSqGSIb3DQEBCwUAA4IBAQA7kxxg2aVjo7Oml83bUWk4UtaQKYMEY74mygG/JV09g1DVMAPAyjaaMFamDSjortKarMQ3ET5tj2DggQBsWQNzsr3iZkmijab8JLwzA2+I1q63S68OaW5uaR5iMR8zZCTh/fWWYqa1AP64XeGHp+RLGfbp/eToNfkQWu7fH2QtDMOeLe5VmIV9pOFHnySszoR/epMd3sdDLVgmz4qbrMTBWD+5rxWdYS2glmRXl7IIQHrdBTRMll7S6ks5prqKFTwfPvZVrTnzD83a39wl2jBJhOQLjmSfSwP9H0YFNb/NRaDbSDS7BPuAlotZsaPZIN95tu+t9wmFwdxcVG/9q/Vu</ds:X509Certificate></ds:X509Data></ds:KeyInfo>"
```
Comment (neongreen) [inline test/XML/Signature.hs:305]: Why was this removed?

File: SAML2/XML/LibXML2.hsc [hunk 8]
```diff
@@ -63,16 +65,55 @@ newtype Doc = Doc{ unDoc :: ForeignPtr XMLDoc }
...
+
+-- | Call libxml2 parser and return the parsed C object, or throw an error.
 fromBytes :: BS.ByteString -> IO Doc
 fromBytes s = do
-  d <- BSU.unsafeUseAsCStringLen s $ \(p, l) ->
-    throwErrnoIfNull "xmlReadMemory" $
-      xmlReadMemory p (fromIntegral l) nullPtr nullPtr (#{const XML_PARSE_NOENT} .|. #{const XML_PARSE_DTDLOAD} .|. #{const XML_PARSE_DTDATTR} .|. #{const XML_PARSE_NONET} .|. #{const XML_PARSE_COMPACT})
-  newDoc d
+  resetErrno
+  let config :: CInt
```
Comment (neongreen) [inline SAML2/XML/LibXML2.hsc:83]: Would be nice to have comments for those flags, because it's not obvious from the names what they do.

File: SAML2/XML/Signature.hs [hunk 9]
```diff
@@ -242,56 +247,115 @@ _verifySignatureOld pks xid doc = do
 -- | take a public key and an xml node ID that points to the sub-tree that needs to be signed, and
 -- return @Right ()@ if it is signed with that key.  otherwise, return a (hopefully helpful) error.
 -- use this if you want to verify signatures, and ignore the rest of this module if you can.
+--
+-- how does this work?:
+--   * dig for the subtree of the input with an ID attribute containing xid (the "signed subtree")
+--   * parse the 'Sigature' subtree in that subtree (we only do envelopped signatures)
```
Comment (neongreen) [inline SAML2/XML/Signature.hs]:
```suggestion
--   * parse the 'Signature' subtree in that subtree (we only do envelopped signatures)
```

**Comment by neongreen** [comment]
I can't really vouch for the logic here. Having more tests for happy and unhappy paths would be nice: good signature, good signature but non-canonical encoding, bad signature, missing signature.

File: SAML2/XML.hs (hunk 6 above)
Comment (fisx) [reply SAML2/XML.hs:108]: yes and yes.

File: test/XML/Signature.hs (hunk 7 above)
Comment (fisx) [reply test/XML/Signature.hs:305]: see commit message.

File: SAML2/XML/LibXML2.hsc (hunk 8 above)
Comment (fisx) [reply SAML2/XML/LibXML2.hsc:83]: it's not my code, and i only touched these to make it slightly easier to read.  also, i would argue this information belongs into the libxml2 manual.

**Comment by fisx** [comment]
> I can't really vouch for the logic here. Having more tests for happy and unhappy paths would be nice: good signature, good signature but non-canonical encoding, bad signature, missing signature.

i agree, but i won't do it in this PR.  note that some of the tests you suggest are in saml2-web-sso (or even in wire-server, i would have to read up on the details).

i will merge this now, if any of the answers i give are not enough please yell and i'll handle that in separate PRs.

**Comment by neongreen** [comment]
Okay
